manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+the point of "perform" and this handler.
+
+The handler uses the "continue" primitive to resume the suspended computation
+with the successor of the offered value. In this example, the computation
+"comp1" performs "Xchg 0" and "Xchg 1", and correspondingly receives the values
+"1" and "2" from the handler. Hence, the whole expression evaluates to "3".
+
+It is useful to note that the we need use locally abstract type "(type a)" in
+the effect handler. The type "Effect.t" is a GADT, and the effect declarations
+may have different type parameters for different effects. The type parameter
+"a" in the type "a Effect.t" represents the type of the value returned when
+performing the effect.
+
+Another point to note is that the catch-all case ``"| _ -> None"'' is necessary
+when handling effects. This case may be intuitively read as ``forward the
+unhandled effects to the outer handler''.
+
+\section{s:effects-concurrency}{Concurrency}
+
+The expressive power of effect handlers comes from the delimited continuation.
+While the previous example immediately resumed the computation, the computation
+may be resumed later, running some other computation in the interim. Let us
+extend the previous example and implement message-passing concurrency between
+two concurrent computations using the "Xchg" effect. We call these concurrent
+computations \textit{tasks}.
+
+A task may either be in suspended state or is completed. We represent the task
+status as follows:
+
+\begin{caml_example*}{verbatim}
+type 'a status =
+  Complete of 'a
+| Suspended of {msg: int; cont: (int, 'a status) continuation}
+\end{caml_example*}
+
+The task may either have been "Complete" with the result or is "Suspended" with
+the message "msg" to send and the continuation "cont". The type "(int, 'a
+status) continuation" says that the suspended computation expects an "int"
+value to resume and returns a "'a status" value when resumed.
+
+Next, we define a "step" function that executes one step of computation until
+it completes or suspends:
+
+\begin{caml_example*}{verbatim}
+let step f () =
+  match_with f ()
+  { retc = (fun v -> Complete v);
+    exnc = raise;
+    effc = fun (type a) (eff: a t) ->
+      match eff with
+      | Xchg msg -> Some (fun (cont: (a, _) continuation) ->
+          Suspended {msg; cont})
+      | _ -> None }
+\end{caml_example*}
+
+Like "try_with", "match_with" also introduces an effect handler. Unlike
+"try_with", where only the effect cases "effc" is provided, "match_with"
+expects the handlers for the value ("retc") and exceptional ("exnc") return
+cases.
+
+\begin{itemize}
+  \item Case "retc": If the computation returns with a value "v", we return
+    "Complete v".
+  \item Case "exnc": If the computation raises an exception, then the handler
+    raises the same exception.
+  \item Case "effc": If the computation performs the effect "Xchg msg" with the
+    continuation "cont", then we return "Suspended{msg;cont}".
```

Comment:
I would suggest adding: "Thus, in this case, the continuation `cont` is not immediately invoked by the handler; instead, it is stored in a data structure for later use."